src/peep/peep.c

```diff
@@ -7772,13 +7772,285 @@ static void peep_on_enter_or_exit_ride(rct_peep *peep, int rideIndex, int flags)
+
+///loc_69B2B7:
+	// The peep has now decided to buy the item (or, specifically, has not been
+	// dissuaded so far).
+	if (shopItem >= 32)
+		peep->item_extra_flags |= (1 << (shopItem >= 32 ? (shopItem - 32) : shopItem));
+	else
+		peep->item_standard_flags |= (1 << (shopItem >= 32 ? (shopItem - 32) : shopItem));
+
+/// loc_69B2C8:
+	if (shopItem == SHOP_ITEM_TSHIRT)
+		peep->tshirt_colour = ride->track_colour_main[0];
+
+/// loc_69B2E0:
+	if (shopItem == SHOP_ITEM_HAT)
+		peep->hat_colour = ride->track_colour_main[0];
+
+/// loc_69B2FB:
+	if (shopItem == SHOP_ITEM_BALLOON)
+		peep->balloon_colour = ride->track_colour_main[0];
+
+/// loc_69B316:
+	if (shopItem == SHOP_ITEM_UMBRELLA)
+		peep->umbrella_colour = ride->track_colour_main[0];
+
+/// loc_69B331:
+	if (shopItem == SHOP_ITEM_MAP)
+		sub_69A98C(peep);
+
+/// loc_69B33A:
+	uint16 dl;
+	if (shopItem >= 32)
+		dl = RCT2_ADDRESS(0x982310, uint8)[shopItem - 32];
+	else
+		dl = RCT2_ADDRESS(0x9822F4, uint8)[shopItem];
+/// loc_69B34C:
+	peep->var_42 = min((peep->var_42 + dl), 255);
+
+/// loc_69B355:
+	if (shopItem == SHOP_ITEM_PHOTO)
+		peep->photo1_ride_ref = rideIndex;
+
+/// loc_69B35F:
+	if (shopItem == SHOP_ITEM_PHOTO2)
+		peep->photo2_ride_ref = rideIndex;
+
+/// loc_69B366:
+	if (shopItem == SHOP_ITEM_PHOTO3)
+		peep->photo3_ride_ref = rideIndex;
+
+/// loc_69B36D:
+	if (shopItem == SHOP_ITEM_PHOTO4)
+		peep->photo4_ride_ref = rideIndex;
+
+/// loc_69B374:
+	peep->window_invalidate_flags |= PEEP_INVALIDATE_PEEP_INVENTORY;
+	peep_update_sprite_type(peep);
+	if (peep->flags & PEEP_FLAGS_TRACKING) {
+		RCT2_GLOBAL(0x13CE952,uint16) = peep->name_string_idx;
+		RCT2_GLOBAL((0x13CE952 + 2), uint32) = peep->id;
+		// loc_69B3B5:
+		RCT2_GLOBAL((0x13CE956 + 2), uint16) = (shopItem >= 32 ? STR_SHOP_ITEM_INDEFINITE_PHOTO2 + (shopItem - 32) : STR_SHOP_ITEM_INDEFINITE_BALLOON + shopItem);
+		news_item_add_to_queue(2, STR_PEEP_TRACKING_NOTIFICATION_BOUGHT_X, peep->sprite_index);
+	}
+
+	if (shop_item_is_food(shopItem))
+		peep->no_of_food++;
+
+	if (shop_item_is_drink(shopItem))
+		peep->no_of_drinks++;
+
+	if (shop_item_is_souvenir(shopItem))
+		peep->no_of_souvenirs++;
+
+	money16* expend_type = &peep->paid_on_souvenirs;
+
+	RCT2_GLOBAL(RCT2_ADDRESS_NEXT_EXPENDITURE_TYPE, uint8) = RCT_EXPENDITURE_TYPE_SHOP_STOCK * 4;
+
+	/// loc_69B3CC:
+	if (shop_item_is_food(shopItem)) {
+		expend_type = &peep->paid_on_food;
+		RCT2_GLOBAL(RCT2_ADDRESS_NEXT_EXPENDITURE_TYPE, uint8) = RCT_EXPENDITURE_TYPE_FOODDRINK_STOCK * 4;
+	}
+
+	/// loc_69B4B1 / loc_69B430:
+	if (shop_item_is_drink(shopItem)) {
+		expend_type = &peep->paid_on_drink;
+		RCT2_GLOBAL(RCT2_ADDRESS_NEXT_EXPENDITURE_TYPE, uint8) = RCT_EXPENDITURE_TYPE_FOODDRINK_STOCK * 4;
+	}
+
+	/// loc_69B4D4:
+	if (!(RCT2_GLOBAL(RCT2_ADDRESS_PARK_FLAGS, uint32) & PARK_FLAGS_NO_MONEY))
+		finance_payment(price, (RCT2_GLOBAL(RCT2_ADDRESS_NEXT_EXPENDITURE_TYPE, uint8) / 4));
+
+	/// loc_69B4E5:
+	RCT2_GLOBAL(RCT2_ADDRESS_NEXT_EXPENDITURE_TYPE, uint8) += 0xFC;
```

Comment:
I went with duncan's idea since it seems to be the best trade-off between readability and compatibility. It's something that can be revisited once all of the side-effects are known.